train_concept_identification.py

```diff
@@ -140,26 +151,67 @@ def extract_padding(outputs, eos_index):
 
 
 def indices_to_words(outputs_no_padding,
-                     vocabs: Vocabs):
-  ids_to_concepts_list = list(vocabs.concept_vocab.keys())
-  concepts_as_list = [ids_to_concepts_list[int(id)]\
-                        for sentence in outputs_no_padding for id in sentence\
-                        if ids_to_concepts_list[int(id)] != UNK]
+                     extended_vocab,
+                     config: CfgNode):
+
+  if config.USE_POINTER_GENERATOR:
+    vocab = extended_vocab
+  else:
+    vocab = extended_vocab.concept_vocab
+
+  ids_to_concepts_list = list(vocab.keys())
+  concepts_as_list = []
+  for sentence in outputs_no_padding:
+    concepts = []
+    for id in sentence:
+        if ids_to_concepts_list[int(id)] != UNK: concepts.append(ids_to_concepts_list[int(id)])
+    concepts_as_list.append(concepts)
   return concepts_as_list
 
 
 def eval_step(model: nn.Module,
               criterion: nn.Module,
               max_out_len: int,
               vocabs: Vocabs,
-              batch: Dict[str, torch.tensor]):
+              batch: Dict[str, torch.tensor],
+              config: CfgNode, device):
   inputs = batch['sentence']
   inputs_lengths = batch['sentence_lengts']
   gold_outputs = batch['concepts']
 
-  logits, predictions = model(inputs, inputs_lengths, max_out_length=max_out_len)
+  if config.USE_POINTER_GENERATOR:
+    unnumericalized_inputs = batch['initial_sentence']
+    unnumericalized_concepts = batch['concepts_string']
+    # compute extended vocab
+    extended_vocab = deepcopy(vocabs.shared_vocab)
+
+     # compute extended vocabulary size
+    extended_vocab_size = len(extended_vocab.items())
+
+    # add in the extended vocabulary the words from the initial input
+    for sentence in unnumericalized_inputs:
+        for token in sentence:
+            if token not in extended_vocab.keys():
+                extended_vocab[token] = extended_vocab_size
+                extended_vocab_size += 1
```

Comment:
It would be best to move this into a method (maybe in the same script, maybe in utils) that returns the extended vocabulary
You can call it 'construct_extended_vocabulary' or something along those lines. It would be best to also be documented